main.cpp

```diff
@@ -21,38 +20,136 @@ int main(int argc, char **argv)
 
         // check validity
         if (filter == '?') {
-            throw std::string {"Invalid filter."};
+            throw string {"Invalid filter."};
         }
 
         // Ensure only one filter
         if (getopt(argc, argv, filtersOptions) != -1) {
-            throw std::string {"Only one filter allowed."};
+            throw string {"Only one filter allowed."};
         }
 
         // Ensure proper usage
         if (argc != optind + 2) {
-            throw std::string {"Usage: filter [flag] infile outfile"};
+            throw string {"Usage: filter [flag] infile outfile"};
         }
 
-        // Open input file and check
-        FILE *inptr {fopen(argv[optind], "r")};
-        if (inptr == nullptr) {
-            throw std::string {"Could not read : " + static_cast<std::string>(argv[optind])};
+        // Open input file 
+        FILE *inFile {fopen(argv[optind], "r")};
+        // check readability
+        if (inFile == nullptr) {
+            throw string {"Could not read : " + static_cast<string>(argv[optind])};
         }
 
         // open output file
-        FILE *outptr {fopen(argv[optind + 1], "w")};
-        if (outptr == nullptr) {
-            throw std::string {"Could not create : " + static_cast<std::string>(argv[optind + 1])};
+        FILE *outFile {fopen(argv[optind + 1], "w")};
+        // check writability
+        if (outFile == nullptr) {
+            throw string {"Could not create : " + static_cast<string>(argv[optind + 1])};
         }
 
+        // Read inFile's BITMAPFILEHEADER
+        BITMAPFILEHEADER bf;
+        fread(&bf, sizeof(BITMAPFILEHEADER), 1, inFile);
+
+        // Read inFile's BITMAPINFOHEADER
+        BITMAPINFOHEADER bi;
+        fread(&bi, sizeof(BITMAPINFOHEADER), 1, inFile);
+
+        // Ensure infile is (likely) a 24-bit uncompressed BMP 4.0
+        if (bf.bfType != 0x4d42 || bf.bfoffBits != 54 || bi.biSize != 40 || bi.biBitCount != 24 || bi.biCompression != 0) {
+            fclose(outFile);
+            fclose(inFile);
+            
+            throw string {"Unsupported file format."};
+        }
+
+        int height {abs(bi.biHeight)};
+        int width {abs(bi.biWidth)};
+
+        // Allocate memory for image
+        RGBTRIPLE *image {new RGBTRIPLE[height * width]};
```

Comment:
```c
RGBTRIPLE(*image)[width] = calloc(height, width * sizeof(RGBTRIPLE));
```
Here `pointer to an array` is used or in short `array pointer`. Is this similar to what we are expecting from the new code? @chandrakant100 @Ank1taS

Reply:
Yes, about this line I actually forgot to mention.
The image must be an array of pointers of size = width pointing to `RGBTRIPLE`.

Reply:
https://www.geeksforgeeks.org/how-to-declare-a-2d-array-dynamically-in-c-using-new-operator/
There are two ways to dynamically create a 2D array.
And @Ank1taS had used the first method.

Reply:
Array of pointer or pointer to an array?

Reply:
Are we expecting 2D array here? Idk. May we need to see it.

Reply:
```suggestion
        RGBTRIPLE **image {new RGBTRIPLE[width]};
        for (int i{0}; i < height; i++)
        {
               image[i] = new RGBTRIPLE[width];
        }
```

Reply:
> 

Yes, Because thereafter we are passing that into out main filter function in switch case(last parameter which is expecting a 2d array).

Reply:
> Are we expecting 2D array here? Idk. May we need to see it.

Yes, Because thereafter we are passing that into out main filter function in switch case(last parameter which is expecting a 2d array).

Reply:
My bad...
I should have to declare an array of ponters. I will update it.